File: ingestion/ingestion/doc_model.py
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, Field, HttpUrl


class TocItem(BaseModel):
    id: str
    title: str
    level: int = Field(ge=1, le=6)


class InlineText(BaseModel):
    type: Literal["text"] = "text"
    text: str


class InlineCode(BaseModel):
    type: Literal["code"] = "code"
    text: str


class InlineEmphasis(BaseModel):
    type: Literal["emphasis"] = "emphasis"
    inlines: list[InlineNode]


class InlineStrong(BaseModel):
    type: Literal["strong"] = "strong"
    inlines: list[InlineNode]


class InlineLink(BaseModel):
    type: Literal["link"] = "link"
    href: str
    inlines: list[InlineNode]
    linkType: Literal["internal", "external", "unresolved"]

    @staticmethod
    def _ensure_allowed_external(url: str) -> None:
        HttpUrl(url)

    @staticmethod
    def _ensure_allowed_internal(url: str) -> None:
        if not url.startswith("/"):
            raise ValueError("internal href must start with '/'")

    def model_post_init(self, __context: object) -> None:
        if self.linkType == "external":
            self._ensure_allowed_external(self.href)
        else:
            self._ensure_allowed_internal(self.href)


InlineNode = Annotated[
    InlineText | InlineCode | InlineEmphasis | InlineStrong | InlineLink,
    Field(discriminator="type"),
]


class BlockHeading(BaseModel):
    type: Literal["heading"] = "heading"
    id: str
    level: int = Field(ge=1, le=6)
    text: str


class BlockParagraph(BaseModel):
    type: Literal["paragraph"] = "paragraph"
    inlines: list[InlineNode]


class BlockList(BaseModel):
    type: Literal["list"] = "list"
    ordered: bool
    items: list[list[BlockNode]]


class DefinitionListItem(BaseModel):
    id: str | None = None
    termInlines: list[InlineNode]
    definitionBlocks: list[BlockNode]


class BlockDefinitionList(BaseModel):
    type: Literal["definition_list"] = "definition_list"
    items: list[DefinitionListItem]


class BlockCode(BaseModel):
    type: Literal["code_block"] = "code_block"
    text: str
    languageHint: str | None = None
    id: str | None = None


class BlockTable(BaseModel):
    type: Literal["table"] = "table"
    headers: list[str]
    rows: list[list[str]]


class BlockHorizontalRule(BaseModel):
    type: Literal["horizontal_rule"] = "horizontal_rule"


BlockNode = Annotated[
    BlockHeading
    | BlockParagraph
    | BlockList
    | BlockDefinitionList
    | BlockCode
    | BlockTable
    | BlockHorizontalRule,
    Field(discriminator="type"),
]


class OptionItem(BaseModel):
    flags: str
    argument: str | None = None
    description: str
    anchorId: str


class SeeAlsoRef(BaseModel):
    name: str
    section: str | None = None
    resolvedPageId: str | None = None


class DocumentModel(BaseModel):
    toc: list[TocItem]
    blocks: list[BlockNode]


# Required for forward references used in union models above.
InlineEmphasis.model_rebuild()
InlineStrong.model_rebuild()
InlineLink.model_rebuild()
BlockList.model_rebuild()
DefinitionListItem.model_rebuild()
BlockDefinitionList.model_rebuild()
DocumentModel.model_rebuild()

File: nextjs/lib/docModel.ts
/**
 * Document model produced by the ingestion pipeline.
 *
 * The Python side is `ingestion/ingestion/doc_model.py`. Both sides are
 * checked against the golden fixtures under `ingestion/tests/fixtures/golden`
 * and `components/doc/__fixtures__`; change them together.
 */

export type TextInline = { type: 'text'; text: string }
export type CodeInline = { type: 'code'; text: string }
export type EmphasisInline = { type: 'emphasis'; inlines: InlineNode[] }
export type StrongInline = { type: 'strong'; inlines: InlineNode[] }
export type LinkInline = {
  type: 'link'
  href: string
  inlines: InlineNode[]
  linkType: 'internal' | 'external' | 'unresolved'
}
export type InlineNode = TextInline | CodeInline | EmphasisInline | StrongInline | LinkInline

export type HeadingBlock = { type: 'heading'; id: string; level: number; text: string }
export type ParagraphBlock = { type: 'paragraph'; inlines: InlineNode[] }
export type ListBlock = { type: 'list'; ordered: boolean; items: BlockNode[][] }
export type DefinitionListItem = {
  id?: string | null
  termInlines: InlineNode[]
  definitionBlocks: BlockNode[]
}
export type DefinitionListBlock = { type: 'definition_list'; items: DefinitionListItem[] }
export type CodeBlock = { type: 'code_block'; id?: string | null; languageHint?: string | null; text: string }
export type TableBlock = { type: 'table'; headers: string[]; rows: string[][] }
export type HorizontalRuleBlock = { type: 'horizontal_rule' }
export type BlockNode =
  | HeadingBlock
  | ParagraphBlock
  | ListBlock
  | DefinitionListBlock
  | CodeBlock
  | TableBlock
  | HorizontalRuleBlock

export type TocItem = { id: string; level: number; title: string }
export type DocumentModel = { toc: TocItem[]; blocks: BlockNode[] }

export type OptionItem = { anchorId: string; flags: string; argument?: string | null; description: string }
export type SeeAlsoRef = { name: string; section?: string | null; resolvedPageId?: string | null }

export type ManPage = {
  id: string
  datasetReleaseId: string
  locale: string
  distro: string
  name: string
  section: string
  title: string
  description: string
  sourcePackage?: string | null
  sourcePackageVersion?: string | null
}

export type ManPageContent = {
  blocks: BlockNode[]
  toc: TocItem[]
  options?: OptionItem[] | null
  seeAlso?: SeeAlsoRef[] | null
  synopsis?: string[] | null
}

export type ManPageVariant = { distro: string; datasetReleaseId: string; contentSha256: string }
export type ManPageResponse = { page: ManPage; content: ManPageContent; variants: ManPageVariant[] }

export type ErrorDetail = { code: string; message: string }
export type ApiErrorEnvelope = { error: ErrorDetail }
export type AmbiguousOption = { section: string; title: string; description: string }
export type AmbiguousPageResponse = { error: ErrorDetail; options: AmbiguousOption[] }
